### PR TITLE
Add speaker support in sof-rt5682 m/c driver

### PR DESCRIPTION
**Pull request by ranj063**
Just a reworked PR with all checkpatch issues fixed. Not tested.

@keqiaozhang @bardliao could I please request for a quick test?

### CODE REVIEW

**Comment by keqiaozhang** [comment]
Verified this PR on WHL-I2S, regression test passed.

File: sound/soc/intel/boards/sof_rt5682.c [hunk 1]
```diff
@@ -412,7 +451,37 @@ static struct snd_soc_dai_link *sof_card_dai_links_create(struct device *dev,
...
+		links[id].num_platforms = ARRAY_SIZE(platform_component);
+		links[id].init = speaker_codec_init,
+		links[id].nonatomic = true;
+		links[id].dpcm_playback = 1;
+		links[id].no_pcm = 1;
+		if (is_legacy_cpu) {
+			links[id].cpu_dai_name = devm_kasprintf(dev, GFP_KERNEL,
+								"ssp%d-port",
+								ssp_amp);
+			if (!links[id].name)
```
Comment (bardliao) [inline sound/soc/intel/boards/sof_rt5682.c]: This looks wrong. Should it be links[id].cpu_dai_name?

File: sound/soc/intel/boards/sof_rt5682.c [hunk 2]
```diff
@@ -412,7 +451,37 @@ static struct snd_soc_dai_link *sof_card_dai_links_create(struct device *dev,
...
+				goto devm_err;
+
+		} else {
+			links[id].cpu_dai_name = devm_kasprintf(dev, GFP_KERNEL,
+								"SSP%d Pin",
+								ssp_amp);
+			if (!links[id].name)
```
Comment (bardliao) [inline sound/soc/intel/boards/sof_rt5682.c]: Also here. Should it be links[id].cpu_dai_name?

File: sound/soc/intel/boards/sof_rt5682.c [hunk 3]
```diff
@@ -448,14 +517,19 @@ static int sof_audio_probe(struct platform_device *pdev)
 		sof_rt5682_quirk);
 
 	dai_links = sof_card_dai_links_create(&pdev->dev, sof_rt5682_quirk &
-					      SOF_RT5682_SSP_MASK,
+					      SOF_RT5682_SSP_CODEC_MASK,
+					      sof_rt5682_quirk &
+					      SOF_RT5682_SSP_AMP_MASK,
```
Comment (bardliao) [inline sound/soc/intel/boards/sof_rt5682.c]: I am not sure if it works. SOF_RT5682_SSP_AMP_MASK is GENMASK(8, 6), So  (sof_rt5682_quirk & SOF_RT5682_SSP_AMP_MASK) will be a number in bit 8 to 6, right? Should we modify it to (sof_rt5682_quirk & SOF_RT5682_SSP_AMP_MASK) >> SOF_RT5682_SSP_AMP_SFT? And #define  SOF_RT5682_SSP_AMP_SFT 6

File: sound/soc/intel/boards/sof_rt5682.c [hunk 4]
```diff
@@ -299,23 +323,36 @@ static struct snd_soc_dai_link_component dmic_component[] = {
...
+				(2 + dmic_num + hdmi_num), GFP_KERNEL);
+	} else {
+		links = devm_kzalloc(dev, sizeof(struct snd_soc_dai_link) *
+				(1 + dmic_num + hdmi_num), GFP_KERNEL);
+	}
 	if (!links)
 		goto devm_err;
 
-	/* SSP */
+	/* HS SSP */
```
Comment (plbossart) [inline sound/soc/intel/boards/sof_rt5682.c]: codec SSP

File: sound/soc/intel/boards/sof_rt5682.c [hunk 5]
```diff
@@ -299,23 +323,36 @@ static struct snd_soc_dai_link_component dmic_component[] = {
...
 	int i, id = 0;
 
-	links = devm_kzalloc(dev, sizeof(struct snd_soc_dai_link) *
-			     (1 + dmic_num + hdmi_num), GFP_KERNEL);
+	if (sof_rt5682_quirk & SOF_SPEAKER_AMP_PRESENT) {
+		links = devm_kzalloc(dev, sizeof(struct snd_soc_dai_link) *
+				(2 + dmic_num + hdmi_num), GFP_KERNEL);
+	} else {
+		links = devm_kzalloc(dev, sizeof(struct snd_soc_dai_link) *
+				(1 + dmic_num + hdmi_num), GFP_KERNEL);
```
Comment (plbossart) [inline sound/soc/intel/boards/sof_rt5682.c]: can we simplify, we have two cases where we compute the number of links in this file. One should be enough, let's try and factor all this.

File: sound/soc/intel/boards/sof_rt5682.c (hunk 1 above)
Comment (plbossart) [reply sound/soc/intel/boards/sof_rt5682.c]: yeah, copy/paste

**Comment by ranj063** [comment]
@plbossart @bardliao all comments addressed now.

@keqiaozhang sorry but need one more round of test please.

**Comment by keqiaozhang** [comment]
> @plbossart @bardliao all comments addressed now.
> 
> @keqiaozhang sorry but need one more round of test please.

@ranj063 
This PR is good to go, no regressions.

File: sound/soc/intel/common/soc-acpi-intel-cnl-match.c [hunk 6]
```diff
@@ -37,6 +44,14 @@ struct snd_soc_acpi_mach snd_soc_acpi_intel_cnl_machines[] = {
 		.sof_fw_filename = "sof-cnl.ri",
 		.sof_tplg_filename = "sof-cml-rt5682.tplg",
 	},
+	{
+		.id = "MX98357A",
+		.drv_name = CML_MACHINE_DRIVER,
```
Comment (plbossart) [inline sound/soc/intel/common/soc-acpi-intel-cnl-match.c]: Better to just use the driver name directly for readability, the define doesn't add much value here.

Comment (ranj063) [reply sound/soc/intel/common/soc-acpi-intel-cnl-match.c]: @plbossart fixed now

**Comment by plbossart** [comment]
merging since the last update was minor and the rest already approved.

**Comment by wenqingfu** [comment]
replaced #754

**Pull request merged**